crates/task-impls/src/quorum_proposal/handlers.rs

```diff
@@ -0,0 +1,332 @@
+/// It forms a Three-Chain, if b'' forms a Two-Chain.
+///
+/// We follow this exact logic to determine if we are able to reach a commit and a decide. A commit
+/// is reached when we have a two chain, and a decide is reached when we have a three chain.
+///
+/// # Example
+/// Suppose we have a decide for view 1, and we then move on to get undecided views 2, 3, and 4. Further,
+/// suppose that our *next* proposal is for view 5, but this leader did not see info for view 4, so the
+/// justify qc of the proposal points to view 3. This is fine, and the undecided chain now becomes
+/// 2-3-5.
+///
+/// Assuming we continue with honest leaders, we then eventually could get a chain like: 2-3-5-6-7-8. This
+/// will prompt a decide event to occur (this code), where the `proposal` is for view 8. Now, since the
+/// lowest value in the 3-chain here would be 5 (excluding 8 since we only walk the parents), we begin at
+/// the first link in the chain, and walk back through all undecided views, making our new anchor view 5,
+/// and out new locked view will be 6.
+///
+/// Upon receipt then of a proposal for view 9, assuming it is valid, this entire process will repeat, and
+/// the anchor view will be set to view 6, with the locked view as view 7.
+async fn visit_leaf_chain<TYPES: NodeType, I: NodeImplementation<TYPES>>(
+    proposal: &QuorumProposal<TYPES>,
+    task_state: &QuorumProposalTaskState<TYPES, I>,
+) -> Result<LeafChainTraversalOutcome<TYPES>> {
+    let proposal_view_number = proposal.get_view_number();
+    let proposal_parent_view_number = proposal.justify_qc.get_view_number();
+    // This is the output return type object whose members will be mutated as we traverse.
+    let mut ret = LeafChainTraversalOutcome::default();
+
+    // Are these views consecutive (1-chain)
+    if proposal_parent_view_number + 1 != proposal_view_number {
+        // Since they aren't we can return early before we do anything else.
+        return Ok(ret);
+    }
+
+    // Unpacking here prevents the need to endlessly call the function. These values don't change during
+    // the execution of this code.
+    let consensus_reader = task_state.consensus.read().await;
+    let validated_state_map = consensus_reader.validated_state_map();
+    let saved_leaves = consensus_reader.saved_leaves();
+    let last_decided_view = consensus_reader.last_decided_view();
+    let saved_payloads = consensus_reader.saved_payloads();
+    let vid_shares = consensus_reader.vid_shares();
+
+    // We are in at least a 1-chain, so we start from here.
+    let mut current_chain_length: usize = 1;
+
+    // Get the state so we can traverse the chain to see if we have a 2 or 3 chain.
+    let walk_start_view_number = proposal_parent_view_number;
+
+    // The next view is the next view we're going to traverse, it is the validated state of the
+    // parent of the proposal.
+    let next_view = validated_state_map
```

Comment:
Can we rename `next_view`, `next_leaf` etc. to `parent_*`, to make it clearer that the view number is smaller than the current proposal?

Reply:
Done.